language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -1,50 +1,163 @@
-import qualified Data.IntMap               as IM
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Quote
 import           Language.PlutusCore.Type
-import           PlutusPrelude
-
-type CloneSt = IM.IntMap (TyNameWithKind ())
-
-cloneEnvAssign :: MonadState CloneSt m
-               => Unique
-               -> TyNameWithKind ()
-               -> m ()
-cloneEnvAssign (Unique i) ty = modify (IM.insert i ty)
-
-cloneType :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneType = flip evalStateT mempty . cloneTypeM
-
-replaceTyName :: (MonadQuote m, MonadState CloneSt m) => TyNameWithKind () -> m (TyNameWithKind ())
-replaceTyName (TyNameWithKind (TyName (Name l' n u))) = do
-    u' <- liftQuote freshUnique
-    let tn = void $ TyNameWithKind (TyName (Name l' n u')) -- TODO use lenses
-    cloneEnvAssign u tn
-    pure tn
-
-cloneTypeM :: (MonadQuote m, MonadState CloneSt m) => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneTypeM (TyFix _ tn ty) = do
-    tn' <- replaceTyName (void tn)
-    TyFix () tn' <$> cloneTypeM ty
-cloneTypeM (TyForall _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyForall () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyLam _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyLam () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyApp _ ty ty') = TyApp () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM (TyFun _ ty ty') = TyFun () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM x@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
-    cloneSt <- get
-    case IM.lookup (unUnique u) cloneSt of
-        Just n  -> pure (TyVar () n)
-        Nothing -> pure (void x)
-cloneTypeM x@TyBuiltin{}    = pure (void x)
-cloneTypeM x@TyInt{}        = pure (void x)
+
+import           Control.Applicative
+import           Control.Monad.Reader
+import           Control.Monad.State
+
+import           Data.Foldable
+import           Data.Functor.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.IntSet               as IS
+import qualified Data.List.NonEmpty        as NE
+import           Data.Maybe                (fromMaybe)
+import           Lens.Micro
+
+type Cloning m = (MonadState CloneSt m, MonadReader CloneWhat m, MonadQuote m)
+
+data CloneSt = CloneSt {
+    csMappingStack :: NE.NonEmpty (IM.IntMap Unique) -- ^ A scoped stack of unique mappings.
+    , csBound      :: IS.IntSet -- ^ The set of bound variables seen so far.
+    }
+
+emptyState :: CloneSt
+emptyState = CloneSt {csMappingStack=mempty NE.:| [], csBound=mempty}
+
+addMapping :: Cloning m => Int -> Unique -> m ()
+addMapping oldU newU = modify $ \(CloneSt (top NE.:| rest) bound) -> CloneSt (IM.insert oldU newU top NE.:| rest) bound
+
+lookupMapping :: Cloning m => Unique -> m (Maybe Unique)
+lookupMapping (Unique u) = do
+    mappings <- gets csMappingStack
+    pure $ foldl' (<|>) Nothing (fmap (IM.lookup u) mappings)
```

Comment:
`foldl' (<|>) Nothing` will traverse the entire list regardless of whether anything is already found or not. `foldr (<|>) Nothing` is properly lazy and is spelled as `asum`. In cases like that I usually define the `asumMap` function (in the style of `foldMap`) and add it to the prelude I'm working with. So that would be

```haskell
pure $ asumMap (IM.lookup u) mappings
```

Reply:
It does need to be a left fold, since we need to check them in the right order, though.

Reply:
I think, `foldl` vs `foldr` is not about order, but is about associativity? With `foldl` you get `(a <|> b) <|> c` while with `foldr` you get `a <|> (b <|> c)`. And the latter one is more efficient.

Reply:
I'm not able to observe any difference between

```haskell
foldl' (<|>) empty [Nothing, Just 1, Nothing, Just 2, Nothing]
```

and

```haskell
foldr (<|>) empty [Nothing, Just 1, Nothing, Just 2, Nothing]
```

Do you have an example?

Reply:
No, you're right. I was just confused!